test/unit/puppetlabs/puppetserver/lockable_pool_test.clj

```diff
@@ -167,54 +207,314 @@
+    (let [pool (create-populated-pool 2)
           instance (.borrowItem pool)]
-      (is (= 0 (.size pool)))
-      (.releaseItem pool instance true)
       (is (= 1 (.size pool)))
+      (.releaseItem pool instance true)
+      (is (= 2 (.size pool)))
       (.lock pool)
-      (is (.isWriteLocked lock))
+      (is (nil? @(future (.borrowItemWithTimeout pool
+                                                 1
+                                                 TimeUnit/MICROSECONDS))))
       (.unlock pool)
-      (is (not (.isWriteLocked  lock))))))
+      (is (not (nil? @(future (.borrowItemWithTimeout pool
+                                                      1
+                                                      TimeUnit/MICROSECONDS)))))
+      (is (true? true)))))
+
+(deftest pool-borrow-blocks-borrow-when-pool-empty
+  (testing "borrow from pool blocks while the pool is empty"
+    (let [pool (create-populated-pool 1)
+          item (.borrowItem pool)
+          borrow-thread-started? (promise)
+          borrow-thread-borrowed? (promise)
+          borrow-thread (future (deliver borrow-thread-started? true)
+                                (let [instance (.borrowItem pool)]
+                                  (deliver borrow-thread-borrowed? true)
+                                  (.releaseItem pool instance)))]
+      @borrow-thread-started?
+      (is (not (realized? borrow-thread-borrowed?)))
+      (.releaseItem pool item)
+      @borrow-thread
+      (is (true? true)))))
+
+(deftest pool-timed-borrows-test
+  (testing "pool borrows with timeout"
+    (let [pool (create-populated-pool 1)
+          item (.borrowItem pool)]
+      (testing "borrow times out and returns nil when pool is empty"
+        (is (nil? (.borrowItemWithTimeout pool 1 TimeUnit/MICROSECONDS))))
+      (.releaseItem pool item)
+      (testing "borrow succeeds when pool is non-empty"
+        (is (identical? item (.borrowItemWithTimeout pool
+                                                     1
+                                                     TimeUnit/MICROSECONDS)))))))
+
+(deftest pool-can-do-blocking-borrow-after-borrow-timed-out
+  (testing "can do blocking borrow from pool after previous borrow timed out"
+    (let [pool (create-populated-pool 1)
+          item (.borrowItem pool)]
+      (is (nil? (.borrowItemWithTimeout pool 1 TimeUnit/MICROSECONDS)))
+      (let [borrow-thread-started? (promise)
+            borrow-thread-borrowed? (promise)
+            borrow-thread (future (deliver borrow-thread-started? true)
+                                  (let [instance (.borrowItem pool)]
+                                    (deliver borrow-thread-borrowed? true)
+                                    (.releaseItem pool instance)))]
+        @borrow-thread-started?
+        (is (not (realized? borrow-thread-borrowed?)))
+        (.releaseItem pool item)
+        @borrow-thread
+        (is (true? true))))))
```

Comment:
Doesn't look like the code changes in the commit where this test came in were related but I'm pretty sure this test was failing earlier on with the code changes on my branch - not on the current master branch.  There were a few tests like this that I added for cases where some of the condition variables weren't being signaled appropriately for waiters when needed, leading to borrow requests being stuck indefinitely.